crates/driver/src/domain/competition/solution/fee.rs

```diff
@@ -138,33 +145,58 @@ impl Fulfillment {
         factor: f64,
     ) -> Result<eth::U256, Error> {
         let surplus = self.surplus_over_reference_price(sell_amount, buy_amount, prices)?;
-        let surplus_in_sell_token = self.surplus_in_sell_token(surplus, prices)?;
-        apply_factor(surplus_in_sell_token, factor)
+        apply_factor(surplus, factor)
     }
 
+    /// Computes the volume based fee in surplus token
+    ///
+    /// The volume is defined as a full sell amount (including fees) conversion
+    /// to the full buy amount (user point of view)
     fn fee_from_volume(&self, prices: ClearingPrices, factor: f64) -> Result<eth::U256, Error> {
         let executed = self.executed().0;
-        let executed_sell_amount = match self.order().side {
+        let executed_in_surplus_token = match self.order().side {
             Side::Buy => {
                 // How much `sell_token` we need to sell to buy `executed` amount of `buy_token`
                 executed
                     .checked_mul(prices.buy)
                     .ok_or(trade::Error::Overflow)?
                     .checked_div(prices.sell)
                     .ok_or(trade::Error::DivisionByZero)?
+                    .checked_add(
+                        // surplus_fee is always expressed in sell token
+                        self.surplus_fee()
+                            .map(|fee| fee.0)
+                            .ok_or(trade::Error::ProtocolFeeOnStaticOrder)?,
+                    )
+                    .ok_or(trade::Error::Overflow)?
             }
-            Side::Sell => executed,
+            Side::Sell => {
+                // How much `buy_token` we get when we sell `executed` amount of `sell_token`
+                executed
+                    .checked_mul(prices.sell)
+                    .ok_or(trade::Error::Overflow)?
+                    .checked_div(prices.buy)
+                    .ok_or(trade::Error::DivisionByZero)?
```

Comment:
I still don't see why this makes sense. Since solvers can trade-off fees and surplus 1:1 I don't understand why we exclude fees here (and only on one order type).

Wouldn't the following two executions be equivalent for the user/protocol but result in very different fees?

User Order: Sell 1 ETH for at least 3500 DAI
1. Solution: Clearing price 35000, execution: pay 0.1 ETH, receive 3500 DAI, fee 0.9ETH => 35k DAI volume
2. Solution: Clearing price 3500, execution: pay 1 ETH, receive 3500 DAI, fee 0 => 3.5k DAI volume

Maybe we shouldn't be using the clearing prices here, but instead the fee-adjusted prices and then add the fee back in both cases?

Reply:
In your example, since it's a sell order, volume based fee is in buy token, so both times volume is 3.5k DAI. 

> Maybe we shouldn't be using the clearing prices here, but instead the fee-adjusted prices and then add the fee back in both cases?

Can't use fee-adjusted prices since those do not exist yet and this code is used to determine the fee-adjusted prices.

I think the asymmetry (adding fee to buy case and not adding to sell case) comes from the fact that the fee is expressed/taken from only one of the sides, namely sell side. 

Similar asymmetry exists in the functions that determine the traded amounts: https://github.com/cowprotocol/services/blob/main/crates/driver/src/domain/competition/solution/trade.rs#L104-L136

So, for sell orders, if we want to know the volume in buy token, we need to use uniform prices. For buy orders, if we want to know the volume in sell token, we need to use custom prices. Since we don't have those, alternative is using uniform prices buy adding fee manually.

Reply:
Oh nice, I think this would actually make things a lot easier to understand. We take the volume in the surplus token so

```
let volume = match self.order().side {
  Side::Buy => self.sell_amount(prices, weth),
  Side::Sell => self.buy_amount(prices, weth),
};
```

makes a lot of sense and is much easier to write. It does require refactoring the different ways of passing prices (full vector vs. just the two prices) but this also makes sense imo to be more coherent across the domain.

Reply:
Refactored